### PR TITLE
[CALCITE-3323] Handle arbitrary/unknown functions that have ordinary syntax

### PR DESCRIPTION
**Pull request by fib-seq**
Added costFactory field and updated set to populate with original properties values

### CODE REVIEW

**Comment by fib-seq** [comment]
@julianhyde these are the fixes to the calcite code that for passing in context properly for connConfig() within PlannerImpl.

**Comment by julianhyde** [comment]
Since CALCITE-3323 is fixed already, can you please create a new JIRA case. The description is "CalciteConnectionConfigImpl.set obliterates previous property values".

The cause is that `new Properties(properties)` doesn't copy the previous value as, say, `new HashMap(map)` does.

Can you add a unit test, say in PlannerTest, that if you call set(x, 1) then set(y, 2) on a CalciteConnectionConfigImpl, the result contains x and y.

File: core/src/main/java/org/apache/calcite/config/CalciteConnectionConfigImpl.java [hunk 1]
```diff
@@ -40,8 +40,13 @@ public CalciteConnectionConfigImpl(Properties properties) {
...
   public CalciteConnectionConfigImpl set(CalciteConnectionProperty property,
       String value) {
-    final Properties properties1 = new Properties(properties);
-    properties1.setProperty(property.camelName(), value);
+    final Properties properties1 = new Properties();
+    this.properties.forEach((connectionProp, connectionValue) -> {
+      properties1.setProperty((String) connectionProp, (String) connectionValue);
+    });
+    if (properties1.getProperty(property.camelName(), null) == null) {
+      properties1.setProperty(property.camelName(), value);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/config/CalciteConnectionConfigImpl.java]: If there is already the property key, we do not override ?

File: core/src/main/java/org/apache/calcite/config/CalciteConnectionConfigImpl.java [hunk 2]
```diff
@@ -40,8 +40,13 @@ public CalciteConnectionConfigImpl(Properties properties) {
   /** Returns a copy of this configuration with one property changed. */
   public CalciteConnectionConfigImpl set(CalciteConnectionProperty property,
       String value) {
-    final Properties properties1 = new Properties(properties);
-    properties1.setProperty(property.camelName(), value);
+    final Properties properties1 = new Properties();
+    this.properties.forEach((connectionProp, connectionValue) -> {
+      properties1.setProperty((String) connectionProp, (String) connectionValue);
+    });
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/config/CalciteConnectionConfigImpl.java]: How about `properties.clone()` ?

File: core/src/main/java/org/apache/calcite/config/CalciteConnectionConfigImpl.java (hunk 1 above)
Comment (jinxing64) [reply core/src/main/java/org/apache/calcite/config/CalciteConnectionConfigImpl.java]: +1

File: core/src/main/java/org/apache/calcite/config/CalciteConnectionConfigImpl.java (hunk 2 above)
Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/config/CalciteConnectionConfigImpl.java]: Yes, `clone()` looks promising.

File: core/src/main/java/org/apache/calcite/config/CalciteConnectionConfigImpl.java (hunk 1 above)
Comment (fib-seq) [reply core/src/main/java/org/apache/calcite/config/CalciteConnectionConfigImpl.java]: I agree, ideally it should override if a key is already present. Part of this hack is due to how `connConfig()` obliterates the `CASE_SENSITIVE` and `CONFORMANCE` defaults that may be passed in through `context`. 

Ideally the flow should be in `connConfig()`: 

- set defaults `CASE_SENSITIVE` and `CONFORMANCE`
- check context for `CalciteConnectionConfig`
- if `CalciteConnectionConfig` context exists, set/override defaults properties within context

vs. 
- getting `CalciteConnectionConfig` from context
- set defaults **but** don't override configs from context

I'm not sure the best approach to get the properties from a CalciteConnectionConfig object. Any suggestions would be appreciated.

**Pull request closed without merging**